server/src/main/java/org/elasticsearch/cluster/metadata/MetadataDeleteIndexService.java

```diff
@@ -93,21 +93,32 @@ public void deleteIndices(final DeleteIndexClusterStateUpdateRequest request) {
     public static ClusterState deleteIndices(ClusterState currentState, Set<Index> indices, Settings settings) {
         final Metadata meta = currentState.metadata();
         final Set<Index> indicesToDelete = new HashSet<>();
-        final Map<Index, DataStream> backingIndices = new HashMap<>();
+        final Map<Index, DataStream> dataStreamIndices = new HashMap<>();
         for (Index index : indices) {
             IndexMetadata im = meta.getIndexSafe(index);
             DataStream parent = meta.getIndicesLookup().get(im.getIndex().getName()).getParentDataStream();
             if (parent != null) {
-                if (parent.getWriteIndex().equals(im.getIndex())) {
+                final Index targetWriteIndex;
+                final boolean isFailureStore;
+                if (parent.isFailureStoreIndex(im.getIndex().getName())) {
+                    targetWriteIndex = parent.getFailureStoreWriteIndex();
+                    isFailureStore = true;
+                } else {
+                    targetWriteIndex = parent.getWriteIndex();
+                    isFailureStore = false;
+                }
+                if (targetWriteIndex != null && targetWriteIndex.equals(im.getIndex())) {
```

Comment:
Fun fact, `String.equals()` checks for the `==` case first since Java String is immutable. I'll change it over to this if you think it's more readable.

Reply:
Yeah but the `==` in `String.equals()` won't help us here, because if we're invoking `String.equals()` inside `Index.equals()`, we already know what the index UUIDs won't be equal either, but `String.equals()` will still do a full string comparison.